README.md
# wechat-emoji-parser

微信表情解析工具，可以实现表情解析和表情发送功能

- 可用于公众号接收用户表情消息，将表情码转换为对应的表情 html 内容
- 还可以获取聊天表情，可以应用于聊天应用的表情

## install（安装）

npm:

```bash
npm install wechat-emoji-parser-lyh -S
```

yarn:

```bash
yarn add wechat-emoji-parser-lyh
```

## How to use（使用）

#### getEmojis(option)：获取表情

```js
import { getEmojis } from 'wechat-emoji-parser'
let emojis = getEmojis({ size: 24 })
/*emojis
[
  {
    cn: "[右哼哼]"
    code: "/:@>"
    style:{
            background: "url(https://res.wx.qq.com/wxdoc/dist/assets/img/emoji-sprite.b5bd1fe0.png) no-repeat"
            background-position: "0px 0px"
            background-size: "599.7px"
            display: "inline-block"
            height: "50px"
            width: "50px"
	}
....
]
*/
```

1. 输入：option: {size: number, emojiSpriteUrl: string}

   | 属性名         | 类型   | 默认值                                                                | 说明                                                                                                                 |
   | -------------- | ------ | --------------------------------------------------------------------- | -------------------------------------------------------------------------------------------------------------------- |
   | size           | number | 64                                                                    | emoji 大小，单位 px                                                                                                  |
   | emojiSpriteUrl | string | https://res.wx.qq.com/wxdoc/dist/assets/img/emoji-sprite.b5bd1fe0.png | 雪碧图文件路径,为了避免链接失效，最好将 src/assets/emoji-sprite.png 中的文件上传自己的 cdn，然后设置为对应的资源路径 |

2. 输出 emojis: [
   {
   code: string,
   cn: string,
   style: {
   ​ display: string
   ​ background: string
   ​ width: string
   ​ height: string
   ​ 'background-position': string
   ​ 'background-size': string
   }

}]

| 属性名 | 类型   | 默认值 | 说明                       |
| ------ | ------ | ------ | -------------------------- |
| code   | string |        | emoji 对应的编码，如：/:@> |
| cn     | string |        | 中文显示，如：[右哼哼]     |
| style  | Object |        |                            |

#### parseEmoji(str) 解析表情

```js
import { parseEmoji, configParseEmoji } from 'wechat-emoji-parser'
configParseEmoji({ size: 30 }) // 设置一些参数
const res = parseEmoji('哈哈[西瓜]') // 解析文本
```

- configParseEmoji(option: object) 配置解析规则

  | 属性名         | 类型   | 默认值                                                                | 说明                                                                                                                 |
  | -------------- | ------ | --------------------------------------------------------------------- | -------------------------------------------------------------------------------------------------------------------- |
  | size           | number | 64                                                                    | emoji 大小，单位 px                                                                                                  |
  | emojiSpriteUrl | string | https://res.wx.qq.com/wxdoc/dist/assets/img/emoji-sprite.b5bd1fe0.png | 雪碧图文件路径,为了避免链接失效，最好将 src/assets/emoji-sprite.png 中的文件上传自己的 cdn，然后设置为对应的资源路径 |
  | tag            | string | 'a'                                                                   | 解析后的 html 标签                                                                                                   |

- parseEmoji(str: string)

  输入：待解析的文本

  输出：解析后的内容

### example

- [vue 项目使用案例](./examples/vue-example)


CHANGELOG.md
## [2.0.3](https://github.com/EHG613/wechat-emoji-parser/compare/v2.0.2...v2.0.3) (2021-04-01)



## [2.0.2](https://github.com/EHG613/wechat-emoji-parser/compare/v2.0.1...v2.0.2) (2021-03-30)


### Features

* 压缩输出的代码 ([7cd06db](https://github.com/EHG613/wechat-emoji-parser/commit/7cd06dbfddc42b2e55080b7bedcef37e323b2434))



## 2.0.1 (2021-03-29)


### Bug Fixes

* add dist dir ([1d60ee3](https://github.com/EHG613/wechat-emoji-parser/commit/1d60ee36c857a46833969852da9584c2200bdf23))
* multiple emoji error ([653aa7f](https://github.com/EHG613/wechat-emoji-parser/commit/653aa7f0921d4f091584f174e822cd680004f2af))


### Features

* add 2.x 解析功能 ([2a7bd94](https://github.com/EHG613/wechat-emoji-parser/commit/2a7bd94b5a382ed74e8d2310e4be6d651ba7f27e))
* add 2.x 解析功能 ([fd344a1](https://github.com/EHG613/wechat-emoji-parser/commit/fd344a1829ad18c3a327c5a64b90b729a76ae078))
* add emoji chars ([72f3fb0](https://github.com/EHG613/wechat-emoji-parser/commit/72f3fb00131af676493949ba4fb30ddafde48077))
* add emoji panel ([464a9aa](https://github.com/EHG613/wechat-emoji-parser/commit/464a9aac666070c2816bdac9ac68406dad74f40a))
* add emoji parser ([e22ec89](https://github.com/EHG613/wechat-emoji-parser/commit/e22ec893a73c4eee85f0e8bf6ea40827c42b149c))
* add example for vue ([3c2f777](https://github.com/EHG613/wechat-emoji-parser/commit/3c2f77781314b1f940595b8f2bc9335b5875f50f))
* add example for vue ([dc4f30f](https://github.com/EHG613/wechat-emoji-parser/commit/dc4f30f39be14a9f653461888092353dc277cf7d))
* add getEmojis ([092d6c5](https://github.com/EHG613/wechat-emoji-parser/commit/092d6c517c387ad321de288aa5b262aa83cb7944))
* add more emoji ([14a409f](https://github.com/EHG613/wechat-emoji-parser/commit/14a409f488cbc4eee29f970e7812ec4fe0eaef44))
* add structor ([9e2e425](https://github.com/EHG613/wechat-emoji-parser/commit/9e2e42533dc9017611d57fc215eb245b25332341))
* 添加更多表情 ([f16451c](https://github.com/EHG613/wechat-emoji-parser/commit/f16451c540a58abe21815d639e59e43834204add))



# 2.0.0 (2021-03-19)


### Bug Fixes

* add dist dir ([1d60ee3](https://github.com/mingtianyihou33/wechat-emoji-parser/commit/1d60ee36c857a46833969852da9584c2200bdf23))
* multiple emoji error ([653aa7f](https://github.com/mingtianyihou33/wechat-emoji-parser/commit/653aa7f0921d4f091584f174e822cd680004f2af))


### Features

* add 2.x 解析功能 ([2a7bd94](https://github.com/mingtianyihou33/wechat-emoji-parser/commit/2a7bd94b5a382ed74e8d2310e4be6d651ba7f27e))
* add 2.x 解析功能 ([fd344a1](https://github.com/mingtianyihou33/wechat-emoji-parser/commit/fd344a1829ad18c3a327c5a64b90b729a76ae078))
* add emoji chars ([72f3fb0](https://github.com/mingtianyihou33/wechat-emoji-parser/commit/72f3fb00131af676493949ba4fb30ddafde48077))
* add emoji panel ([464a9aa](https://github.com/mingtianyihou33/wechat-emoji-parser/commit/464a9aac666070c2816bdac9ac68406dad74f40a))
* add emoji parser ([e22ec89](https://github.com/mingtianyihou33/wechat-emoji-parser/commit/e22ec893a73c4eee85f0e8bf6ea40827c42b149c))
* add example for vue ([3c2f777](https://github.com/mingtianyihou33/wechat-emoji-parser/commit/3c2f77781314b1f940595b8f2bc9335b5875f50f))
* add example for vue ([dc4f30f](https://github.com/mingtianyihou33/wechat-emoji-parser/commit/dc4f30f39be14a9f653461888092353dc277cf7d))
* add getEmojis ([092d6c5](https://github.com/mingtianyihou33/wechat-emoji-parser/commit/092d6c517c387ad321de288aa5b262aa83cb7944))
* add more emoji ([14a409f](https://github.com/mingtianyihou33/wechat-emoji-parser/commit/14a409f488cbc4eee29f970e7812ec4fe0eaef44))
* add structor ([9e2e425](https://github.com/mingtianyihou33/wechat-emoji-parser/commit/9e2e42533dc9017611d57fc215eb245b25332341))
* 添加更多表情 ([f16451c](https://github.com/mingtianyihou33/wechat-emoji-parser/commit/f16451c540a58abe21815d639e59e43834204add))



## 1.1.1 (2020-07-01)

### Bug Fixes

- add dist dir ([1d60ee3](https://github.com/mingtianyihou33/wechat-emoji-parser/commit/1d60ee36c857a46833969852da9584c2200bdf23))
- multiple emoji error ([653aa7f](https://github.com/mingtianyihou33/wechat-emoji-parser/commit/653aa7f0921d4f091584f174e822cd680004f2af))

### Features

- add emoji chars ([72f3fb0](https://github.com/mingtianyihou33/wechat-emoji-parser/commit/72f3fb00131af676493949ba4fb30ddafde48077))
- add emoji panel ([464a9aa](https://github.com/mingtianyihou33/wechat-emoji-parser/commit/464a9aac666070c2816bdac9ac68406dad74f40a))
- add emoji parser ([e22ec89](https://github.com/mingtianyihou33/wechat-emoji-parser/commit/e22ec893a73c4eee85f0e8bf6ea40827c42b149c))
- add more emoji ([14a409f](https://github.com/mingtianyihou33/wechat-emoji-parser/commit/14a409f488cbc4eee29f970e7812ec4fe0eaef44))
- add structor ([9e2e425](https://github.com/mingtianyihou33/wechat-emoji-parser/commit/9e2e42533dc9017611d57fc215eb245b25332341))
- 添加更多表情 ([f16451c](https://github.com/mingtianyihou33/wechat-emoji-parser/commit/f16451c540a58abe21815d639e59e43834204add))


src/config/emoji.ts
export const EMOJI_PANEL: EmojiPanel = {
  width: 724,
  height: 658,
  x: 11,
  y: 10,
  paddingTop: 0,
  paddingBottom: 0,
  paddingLeft: 0,
  paddingRight: 0,
  gapX: 6,
  gapY: 6,
}
export const EMOJI_DATA: EmojiData[] = [
  {
    position: { x: 0, y: 0 },
    id: 0,
    cn: '[右哼哼]',
    hk: '[右哼哼]',
    us: '[Bah！R]',
    code: '/:@>',
    web_code: '/右哼哼',
  },
  {
    position: { x: 1, y: 0 },
    id: 1,
    cn: '[奸笑]',
    hk: '[奸笑]',
    us: '[Smirk]',
    code: '',
    web_code: '',
  },
  {
    position: { x: 2, y: 0 },
    id: 2,
    cn: '[红包]',
    hk: '[Packet]',
    us: '[Packet]',
    code: '',
    web_code: '',
  },
  {
    position: { x: 3, y: 0 },
    id: 3,
    cn: '[生病]',
    emoji: '😷',
    hk: '',
    us: '[Sick]',
    code: '\\ue40c',
    web_code: '',
  },
  {
    position: { x: 4, y: 0 },
    id: 4,
    cn: '[冷汗]',
    hk: '[冷汗]',
    us: '[Blush]',
    code: '/:--b',
    web_code: '/冷汗',
  },
  {
    position: { x: 5, y: 0 },
    id: 5,
    cn: '[惊恐]',
    hk: '[驚恐]',
    us: '[Panic]',
    code: '/::!',
    web_code: '/惊恐',
  },
  {
    position: { x: 6, y: 0 },
    id: 6,
    cn: '[Emm]',
    hk: '[]',
    us: '[]',
    code: '',
    web_code: '',
  },
  {
    position: { x: 7, y: 0 },
    id: 7,
    cn: '[得意]',
    hk: '[得意]',
    us: '[CoolGuy]',
    code: '/:8-)',
    web_code: '/得意',
  },
  {
    position: { x: 8, y: 0 },
    id: 8,
    cn: '[菜刀]',
    hk: '[菜刀]',
    us: '[Cleaver]',
    code: '/:pd',
    web_code: '/菜刀',
  },
  {
    position: { x: 9, y: 0 },
    id: 9,
    cn: '[太阳]',
    hk: '[太陽]',
    us: '[Sun]',
    code: '/:sun',
    web_code: '/太阳',
  },
  {
    position: { x: 10, y: 0 },
    id: 10,
    cn: '[强壮]',
    emoji: '💪',
    hk: '',
    us: '',
    code: '\\ue14c',
    web_code: '',
  },
  {
    position: { x: 0, y: 1 },
    id: 11,
    cn: '[捂脸]',
    hk: '[掩面]',
    us: '[Facepalm]',
    code: '',
    web_code: '',
  },
  {
    position: { x: 1, y: 1 },
    id: 12,
    cn: '[机智]',
    hk: '[機智]',
    us: '[Smart]',
    code: '',
    web_code: '',
  },
  {
    position: { x: 2, y: 1 },
    id: 13,
    cn: '[耶]',
    hk: '[歐耶]',
    us: '[Yeah!]',
    code: '',
    web_code: '',
  },
  {
    position: { x: 3, y: 1 },
    id: 14,
    cn: '[尴尬]',
    hk: '[尷尬]',
    us: '[Awkward]',
    code: '/::-|',
    web_code: '/尴尬',
  },
  {
    position: { x: 4, y: 1 },
    id: 15,
    cn: '[抓狂]',
    hk: '[抓狂]',
    us: '[Scream]',
    code: '/::Q',
    web_code: '/抓狂',
  },
  {
    position: { x: 5, y: 1 },
    id: 16,
    cn: '[流汗]',
    hk: '[流汗]',
    us: '[Sweat]',
    code: '/::L',
    web_code: '/流汗',
  },
  {
    position: { x: 6, y: 1 },
    id: 17,
    cn: '[社会社会]',
    hk: '[]',
    us: '[Respect]',
    code: '',
    web_code: '',
  },
  {
    position: { x: 7, y: 1 },
    id: 18,
    cn: '[擦汗]',
    hk: '[擦汗]',
    us: '[Speechless]',
    code: '/:wipe',
    web_code: '/擦汗',
  },
  {
    position: { x: 8, y: 1 },
    id: 19,
    cn: '[西瓜]',
    hk: '[西瓜]',
    us: '[Watermelon]',
    code: '/:<W>',
    web_code: '/西瓜',
  },
  {
    position: { x: 9, y: 1 },
    id: 20,
    cn: '[拥抱]',
    hk: '[擁抱]',
    us: '[Hug]',
    code: '/:hug',
    web_code: '/拥抱',
  },
  {
    position: { x: 10, y: 1 },
    id: 21,
    cn: '[破涕为笑]',
    emoji: '😂',
    hk: '',
    us: '[Lol]',
    code: '\\ue412',
    web_code: '',
  },
  {
    position: { x: 0, y: 2 },
    id: 22,
    cn: '[皱眉]',
    hk: '[皺眉]',
    us: '[Concerned]',
    code: '',
    web_code: '',
  },
  {
    position: { x: 1, y: 2 },
    id: 23,
    cn: '[鸡]',
    hk: '[小雞]',
    us: '[Chick]',
    code: '',
    web_code: '',
  },
  {
    position: { x: 2, y: 2 },
    id: 24,
    cn: '[微笑]',
    hk: '[微笑]',
    us: '[Smile]',
    code: '/::)',
    web_code: '/微笑',
  },
  {
    position: { x: 3, y: 2 },
    id: 25,
    cn: '[发怒]',
    hk: '[發怒]',
    us: '[Angry]',
    code: '/::@',
    web_code: '/发怒',
  },
  {
    position: { x: 4, y: 2 },
    id: 26,
    cn: '[吐]',
    hk: '[吐]',
    us: '[Puke]',
    code: '/::T',
    web_code: '/吐',
  },
  {
    position: { x: 5, y: 2 },
    id: 27,
    cn: '[憨笑]',
    hk: '[大笑]',
    us: '[Laugh]',
    code: '/::>',
    web_code: '/憨笑',
  },
  {
    position: { x: 6, y: 2 },
    id: 28,
    cn: '[旺柴]',
    hk: '[]',
    us: '[Doge]',
    code: '',
    web_code: '',
  },
  {
    position: { x: 7, y: 2 },
    id: 29,
    cn: '[抠鼻]',
    hk: '[摳鼻]',
    us: '[NosePick]',
    code: '/:dig',
    web_code: '/抠鼻',
  },
  {
    position: { x: 8, y: 2 },
    id: 30,
    cn: '[啤酒]',
    hk: '[啤酒]',
    us: '[Beer]',
    code: '/:beer',
    web_code: '/啤酒',
  },
  {
    position: { x: 9, y: 2 },
    id: 31,
    cn: '[强]',
    hk: '[強]',
    us: '[ThumbsUp]',
    code: '/:strong',
    web_code: '/强',
  },
  {
    position: { x: 10, y: 2 },
    id: 32,
    cn: '[笑脸]',
    emoji: '😄',
    hk: '',
    us: '[Happy]',
    code: '\\ue415',
    web_code: '',
  },
  {
    position: { x: 0, y: 3 },
    id: 33,
    cn: '[调皮]',
    hk: '[調皮]',
    us: '[Tongue]',
    code: '/::P',
    web_code: '/调皮',
  },
  {
    position: { x: 1, y: 3 },
    id: 34,
    cn: '[呲牙]',
    hk: '[呲牙]',
    us: '[Grin]',
    code: '/::D',
    web_code: '/呲牙',
  },
  {
    position: { x: 2, y: 3 },
    id: 35,
    cn: '[惊讶]',
    hk: '[驚訝]',
    us: '[Surprise]',
    code: '/::O',
    web_code: '/惊讶',
  },
  {
    position: { x: 3, y: 3 },
    id: 36,
    cn: '[难过]',
    hk: '[難過]',
    us: '[Frown]',
    code: '/::(',
    web_code: '/难过',
  },
  {
    position: { x: 4, y: 3 },
    id: 37,
    cn: '[色]',
    hk: '[色]',
    us: '[Drool]',
    code: '/::B',
    web_code: '/色',
  },
  {
    position: { x: 5, y: 3 },
    id: 38,
    cn: '[悠闲]',
    hk: '[悠閑]',
    us: '[Commando]',
    code: '/::,@',
    web_code: '/大兵',
  },
  {
    position: { x: 6, y: 3 },
    id: 39,
    cn: '[疑问]',
    hk: '[疑問]',
    us: '[Shocked]',
    code: '/:?',
    web_code: '/疑问',
  },
  {
    position: { x: 7, y: 3 },
    id: 40,
    cn: '[鼓掌]',
    hk: '[鼓掌]',
    us: '[Clap]',
    code: '/:handclap',
    web_code: '/鼓掌',
  },
  {
    position: { x: 8, y: 3 },
    id: 41,
    cn: '[害羞]',
    hk: '[害羞]',
    us: '[Shy]',
    code: '/::$',
    web_code: '/害羞',
  },
  {
    position: { x: 9, y: 3 },
    id: 42,
    cn: '[睡]',
    hk: '[睡]',
    us: '[Sleep]',
    code: '/::Z',
    web_code: '/睡',
  },
  {
    position: { x: 10, y: 3 },
    id: 43,
    cn: '[无语]',
    emoji: '😒',
    hk: '',
    us: '[Duh]',
    code: '\\ue40e',
    web_code: '',
  },
  {
    position: { x: 0, y: 4 },
    id: 44,
    cn: '[偷笑]',
    hk: '[偷笑]',
    us: '[Chuckle]',
    code: '/:,@P',
    web_code: '/偷笑',
  },
  {
    position: { x: 1, y: 4 },
    id: 45,
    cn: '[愉快]',
    hk: '[愉快]',
    us: '[Joyful]',
    code: '/:,@-D',
    web_code: '/可爱',
  },
  {
    position: { x: 2, y: 4 },
    id: 46,
    cn: '[白眼]',
    hk: '[白眼]',
    us: '[Slight]',
    code: '/::d',
    web_code: '/白眼',
  },
  {
    position: { x: 3, y: 4 },
    id: 47,
    cn: '[傲慢]',
    hk: '[傲慢]',
    us: '[Smug]',
    code: '/:,@o',
    web_code: '/傲慢',
  },
  {
    position: { x: 4, y: 4 },
    id: 48,
    cn: '[困]',
    hk: '[累]',
    us: '[Drowsy]',
    code: '/:|-)',
    web_code: '/困',
  },
  {
    position: { x: 5, y: 4 },
    id: 49,
    cn: '[发呆]',
    hk: '[發呆]',
    us: '[Scowl]',
    code: '/::|',
    web_code: '/发呆',
  },
  {
    position: { x: 6, y: 4 },
    id: 50,
    cn: '[好的]',
    hk: '[]',
    us: '[NoProb]',
    code: '',
    web_code: '',
  },
  {
    position: { x: 7, y: 4 },
    id: 51,
    cn: '[坏笑]',
    hk: '[壞笑]',
    us: '[Trick]',
    code: '/:B-)',
    web_code: '/坏笑',
  },
  {
    position: { x: 8, y: 4 },
    id: 52,
    cn: '[咖啡]',
    hk: '[咖啡]',
    us: '[Coffee]',
    code: '/:coffee',
    web_code: '/咖啡',
  },
  {
    position: { x: 9, y: 4 },
    id: 53,
    cn: '[弱]',
    hk: '[弱]',
    us: '[ThumbsDown]',
    code: '/:weak',
    web_code: '/弱',
  },
  {
    position: { x: 10, y: 4 },
    id: 54,
    cn: '[失望]',
    emoji: '😔',
    hk: '',
    us: '[LetDown]',
    code: '\\ue403',
    web_code: '',
  },
  {
    position: { x: 0, y: 5 },
    id: 55,
    cn: '[奋斗]',
    hk: '[奮鬥]',
    us: '[Determined]',
    code: '/:,@f',
    web_code: '/奋斗',
  },
  {
    position: { x: 1, y: 5 },
    id: 56,
    cn: '[咒骂]',
    hk: '[咒罵]',
    us: '[Scold]',
    code: '/::-S',
    web_code: '/咒骂',
  },
  {
    position: { x: 2, y: 5 },
    id: 57,
    cn: '[吃瓜]',
    hk: '[]',
    us: '[Onlooker]',
    code: '',
    web_code: '',
  },
  {
    position: { x: 3, y: 5 },
    id: 58,
    cn: '[加油]',
    hk: '[]',
    us: '[GoForIt]',
    code: '',
    web_code: '',
  },
  {
    position: { x: 4, y: 5 },
    id: 59,
    cn: '[汗]',
    hk: '[]',
    us: '[Sweats]',
    code: '',
    web_code: '',
  },
  {
    position: { x: 5, y: 5 },
    id: 60,
    cn: '[天啊]',
    hk: '[]',
    us: '[OMG]',
    code: '',
    web_code: '',
  },
  {
    position: { x: 6, y: 5 },
    id: 61,
    cn: '[打脸]',
    hk: '[]',
    us: '[MyBad]',
    code: '',
    web_code: '',
  },
  {
    position: { x: 7, y: 5 },
    id: 62,
    cn: '[左哼哼]',
    hk: '[左哼哼]',
    us: '[Bah！L]',
    code: '/:<@',
    web_code: '/左哼哼',
  },
  {
    position: { x: 8, y: 5 },
    id: 63,
    cn: '[饭]',
    hk: '[飯]',
    us: '[Rice]',
    code: '/:eat',
    web_code: '/饭',
  },
  {
    position: { x: 9, y: 5 },
    id: 64,
    cn: '[握手]',
    hk: '[握手]',
    us: '[Shake]',
    code: '/:share',
    web_code: '/握手',
  },
  {
    position: { x: 10, y: 5 },
    id: 65,
    cn: '[吐舌]',
    emoji: '😝',
    hk: '',
    us: '',
    code: '\\ue409',
    web_code: '',
  },
  {
    position: { x: 0, y: 6 },
    id: 66,
    cn: '[哇]',
    hk: '[]',
    us: '[Wow]',
    code: '',
    web_code: '',
  },
  {
    position: { x: 1, y: 6 },
    id: 67,
    cn: '[嘘]',
    hk: '[噓]',
    us: '[Shhh]',
    code: '/:,@x',
    web_code: '/嘘',
  },
  {
    position: { x: 2, y: 6 },
    id: 68,
    cn: '[晕]',
    hk: '[暈]',
    us: '[Dizzy]',
    code: '/:,@@',
    web_code: '/晕',
  },
  {
    position: { x: 3, y: 6 },
    id: 69,
    cn: '[衰]',
    hk: '[衰]',
    us: '[Toasted]',
    code: '/:,@!',
    web_code: '/衰',
  },
  {
    position: { x: 4, y: 6 },
    id: 70,
    cn: '[骷髅]',
    hk: '[骷髏頭]',
    us: '[Skull]',
    code: '/:!!!',
    web_code: '/骷髅',
  },
  {
    position: { x: 5, y: 6 },
    id: 71,
    cn: '[敲打]',
    hk: '[敲打]',
    us: '[Hammer]',
    code: '/:xx',
    web_code: '/敲打',
  },
  {
    position: { x: 6, y: 6 },
    id: 72,
    cn: '[再见]',
    hk: '[再見]',
    us: '[Wave]',
    code: '/:bye',
    web_code: '/再见',
  },
  {
    position: { x: 7, y: 6 },
    id: 73,
    cn: '[嘿哈]',
    hk: '[吼嘿]',
    us: '[Hey]',
    code: '',
    web_code: '',
  },
  {
    position: { x: 8, y: 6 },
    id: 74,
    cn: '[猪头]',
    hk: '[豬頭]',
    us: '[Pig]',
    code: '/:pig',
    web_code: '/猪头',
  },
  {
    position: { x: 9, y: 6 },
    id: 75,
    cn: '[胜利]',
    hk: '[勝利]',
    us: '[Peace]',
    code: '/:v',
    web_code: '/胜利',
  },
  {
    position: { x: 10, y: 6 },
    id: 76,
    cn: '[恐惧]',
    emoji: '😱',
    hk: '',
    us: '[Terror]',
    code: '\\ue107',
    web_code: '',
  },
  {
    position: { x: 0, y: 7 },
    id: 77,
    cn: '[哈欠]',
    hk: '[哈欠]',
    us: '[Yawn]',
    code: '/::-O',
    web_code: '/哈欠',
  },
  {
    position: { x: 1, y: 7 },
    id: 78,
    cn: '[鄙视]',
    hk: '[鄙視]',
    us: '[Pooh-pooh]',
    code: '/:>-|',
    web_code: '/鄙视',
  },
  {
    position: { x: 2, y: 7 },
    id: 79,
    cn: '[委屈]',
    hk: '[委屈]',
    us: '[Shrunken]',
    code: '/:P-(',
    web_code: '/委屈',
  },
  {
    position: { x: 3, y: 7 },
    id: 80,
    cn: '[流泪]',
    hk: '[流淚]',
    us: '[Sob]',
    code: '/::<',
    web_code: '/流泪',
  },
  {
    position: { x: 4, y: 7 },
    id: 81,
    cn: '[快哭了]',
    hk: '[快哭了]',
    us: '[TearingUp]',
    code: '/::\'|',
    web_code: '/快哭了',
  },
  {
    position: { x: 5, y: 7 },
    id: 82,
    cn: '[阴险]',
    hk: '[陰險]',
    us: '[Sly]',
    code: '/:X-)',
    web_code: '/阴险',
  },
  {
    position: { x: 6, y: 7 },
    id: 83,
    cn: '[亲亲]',
    hk: '[親親]',
    us: '[Kiss]',
    code: '/::*',
    web_code: '/亲亲',
  },
  {
    position: { x: 7, y: 7 },
    id: 84,
    cn: '[可怜]',
    hk: '[可憐]',
    us: '[Whimper]',
    code: '/:8*',
    web_code: '/可怜',
  },
  {
    position: { x: 8, y: 7 },
    id: 85,
    cn: '[玫瑰]',
    hk: '[玫瑰]',
    us: '[Rose]',
    code: '/:rose',
    web_code: '/玫瑰',
  },
  {
    position: { x: 9, y: 7 },
    id: 86,
    cn: '[抱拳]',
    hk: '[抱拳]',
    us: '[Fight]',
    code: '/:@)',
    web_code: '/抱拳',
  },
  {
    position: { x: 10, y: 7 },
    id: 87,
    cn: '[脸红]',
    emoji: '😳',
    hk: '',
    us: '[Flushed]',
    code: '\\ue40d',
    web_code: '',
  },
  {
    position: { x: 0, y: 8 },
    id: 88,
    cn: '[凋谢]',
    hk: '[枯萎]',
    us: '[Wilt]',
    code: '/:fade',
    web_code: '/凋谢',
  },
  {
    position: { x: 1, y: 8 },
    id: 89,
    cn: '[嘴唇]',
    hk: '[嘴唇]',
    us: '[Lips]',
    code: '/:showlove',
    web_code: '/示爱',
  },
  {
    position: { x: 2, y: 8 },
    id: 90,
    cn: '[爱心]',
    hk: '[愛心]',
    us: '[Heart]',
    code: '/:heart',
    web_code: '/爱心',
  },
  {
    position: { x: 3, y: 8 },
    id: 91,
    cn: '[心碎]',
    hk: '[心碎]',
    us: '[BrokenHeart]',
    code: '/:break',
    web_code: '/心碎',
  },
  {
    position: { x: 4, y: 8 },
    id: 92,
    cn: '[蛋糕]',
    hk: '[蛋糕]',
    us: '[Cake]',
    code: '/:cake',
    web_code: '/蛋糕',
  },
  {
    position: { x: 5, y: 8 },
    id: 93,
    cn: '[闭嘴]',
    hk: '[閉嘴]',
    us: '[Silent]',
    code: '/::X',
    web_code: '/闭嘴',
  },
  {
    position: { x: 6, y: 8 },
    id: 94,
    cn: '[炸弹]',
    hk: '[炸彈]',
    us: '[Bomb]',
    code: '/:bome',
    web_code: '/炸弹',
  },
  {
    position: { x: 7, y: 8 },
    id: 95,
    cn: '[便便]',
    hk: '[便便]',
    us: '[Poop]',
    code: '/:shit',
    web_code: '/便便',
  },
  {
    position: { x: 8, y: 8 },
    id: 96,
    cn: '[月亮]',
    hk: '[月亮]',
    us: '[Moon]',
    code: '/:moon',
    web_code: '/月亮',
  },
  {
    position: { x: 9, y: 8 },
    id: 97,
    cn: '[勾引]',
    hk: '[勾引]',
    us: '[Beckon]',
    code: '/:jj',
    web_code: '/勾引',
  },
  {
    position: { x: 10, y: 8 },
    id: 98,
    cn: '[合十]',
    emoji: '🙏',
    hk: '',
    us: '[Worship]',
    code: '\\ue41d',
    web_code: '',
  },
  {
    position: { x: 0, y: 9 },
    id: 99,
    cn: '[拳头]',
    hk: '[拳頭]',
    us: '[Fist]',
    code: '/:@@',
    web_code: '/拳头',
  },
  {
    position: { x: 1, y: 9 },
    id: 100,
    cn: '[OK]',
    hk: '[OK]',
    us: '[OK]',
    code: '/:ok',
    web_code: '/OK',
  },
  {
    position: { x: 2, y: 9 },
    id: 101,
    cn: '[大哭]',
    hk: '[大哭]',
    us: '[Cry]',
    code: '/::\'(',
    web_code: '/大哭',
  },
  {
    position: { x: 3, y: 9 },
    id: 102,
    cn: '[跳跳]',
    hk: '[跳跳]',
    us: '[Waddle]',
    code: '/:jump',
    web_code: '/跳跳',
  },
  {
    position: { x: 4, y: 9 },
    id: 103,
    cn: '[发抖]',
    hk: '[發抖]',
    us: '[Tremble]',
    code: '/:shake',
    web_code: '/发抖',
  },
  {
    position: { x: 5, y: 9 },
    id: 104,
    cn: '[怄火]',
    hk: '[噴火]',
    us: '[Aaagh!]',
    code: '/:<O>',
    web_code: '/怄火',
  },
  {
    position: { x: 6, y: 9 },
    id: 105,
    cn: '[转圈]',
    hk: '[轉圈]',
    us: '[Twirl]',
    code: '/:circle',
    web_code: '/转圈',
  },
  {
    position: { x: 7, y: 9 },
    id: 106,
    cn: '[礼物]',
    hk: '',
    us: '',
    code: '/:gift',
    web_code: '',
  },
  {
    position: { x: 8, y: 9 },
    id: 107,
    cn: '[庆祝]',
    emoji: '🎉',
    hk: '',
    us: '[Party]',
    code: '\\ue312',
    web_code: '',
  },
  {
    position: { x: 9, y: 9 },
    id: 108,
    cn: '[鬼魂]',
    emoji: '👻',
    hk: '',
    us: '',
    code: '\\ue11b',
    web_code: '',
  },
  {
    position: { x: 10, y: 9 },
    id: 109,
    cn: '[撇嘴]',
    hk: '[撇嘴]',
    us: '[Grimace]',
    code: '/::~',
    web_code: '/撇嘴',
  },
]

function transform2KeyPosition(
  data: EmojiData[],
): { [key in string]: EmojiData['position'] } {
  let usedKeys = ['cn', 'us', 'code', 'web_code']
  let res: { [key in string]: EmojiData['position'] } = {}
  data.forEach((item: EmojiData) => {
    usedKeys.forEach((key: keyof Omit<EmojiData, 'id' | 'position'>) => {
      let trieKey = item[key]
      if (trieKey && trieKey !== '[]' && !(trieKey in res)) {
        res[trieKey] = item.position
      }
    })
  })
  return res
}

export const EMOJI_KEY_POSITION = transform2KeyPosition(EMOJI_DATA)
